drivers/lock_queue.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (C) Cloud Software Group, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published
+# by the Free Software Foundation; version 2.1 only.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+#
+# Miscellaneous utility functions
+#
+import os
+import pickle
+import sys
+import time
+
+import lock
+import util
+
+
+DEBUG_LOG = True
+
+def debug_log(msg):
+    if DEBUG_LOG:
+        util.SMlog("LockQueue: " + msg)
+
+def get_process_start_time(pid):
+    proc_file = f"/proc/{pid}/stat"
+    with open(proc_file, 'r') as f:
+        return f.read().split(')')[-1].split(' ')[20]
+
+def process_is_valid(pid, start_time):
+    proc_file = f"/proc/{pid}/stat"
+
+    try:
+        if start_time != get_process_start_time(pid):
+            debug_log(f"Process {pid} has incorrect start time real:{get_process_start_time(pid)} vs expected:{start_time}")
+            return False
+    except FileNotFoundError:
+        debug_log(f"Process {pid} is dead")
+        return False
+
+    return True
+
+class LockQueue:
+    def __init__(self, name):
+        self.name = name
+        self._queue_lock = lock.Lock(name, f"ql-{name}")
+        self._action_lock = lock.Lock(name, f"al-{name}")
+
+        # Open file to hold the process queue
+        self._mem = f"/tmp/mem-{name}"
+        if not os.path.exists(self._mem):
```

Comment:
Surely there is a race condition here. Two processes might simultaneously find that the queue file doesn't exist, and both try to initialize it. So one might set it to be empty, even this were after the other had enqueued itself.